src/System.Management.Automation/engine/runtime/Operations/ClassOps.cs

```diff
@@ -91,31 +101,122 @@ public class ScriptBlockMemberMethodWrapper
+        /// Initialization happens when the script that defines PowerShell class is executed.
+        /// This initialization is required only if this wrapper is for a static method.
+        /// </summary>
+        /// <remakr>
+        /// When the same script file gets executed multiple times, the .NET type generated from the PowerShell class
+        /// defined in the file will be shared in those executions, and thus this method will be called multiple times
+        /// possibly in the contexts of different Runspace/SessionState.
+        ///
+        /// We always use the SessionState from the most recent execution as the default SessionState, so be noted that
+        /// the default SessionState may change over time.
+        ///
+        /// This should be OK because the common usage is to run the static method in the same Runspace where the class
+        /// is declared, and thus we can always get the correct SessionState to use by querying the 'SessionStateKeeper'.
+        /// The default SessionState is used only if a static method is called from a Runspace where the class is never
+        /// defined, or is called on a thread without a default Runspace.
+        /// </remakr>
         internal void InitAtRuntime()
         {
-            var context = Runspace.DefaultRunspace.ExecutionContext;
-            _scriptBlock.Value.SessionStateInternal = context.EngineSessionState;
+            if (_isStatic)
+            {
+                // WeakReference<T>'s instance methods are not thread-safe, so we need the lock to gurantee 
+                // 'SetTarget' and 'TryGetTarget' are not called by multiple threads at the same time.
+                lock (_defaultSessionStateToUse)
+                {
+                    var context = Runspace.DefaultRunspace.ExecutionContext;
+                    _defaultSessionStateToUse.SetTarget(context.EngineSessionState);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Set the SessionState of the script block appropriately.
+        /// </summary>
+        private void PrepareScriptBlockToInvoke(object instance, object sessionStateInternal, object[] args)
+        {
+            SessionStateInternal sessionStateToUse = null;
+            if (instance != null)
+            {
+                // Use the SessionState passed in, which is the one associated with the instance.
+                sessionStateToUse = (SessionStateInternal)sessionStateInternal;
+            }
+            else
+            {
+                // For static method, it's a little complex.
+                // - Check if the current default runspace is registered with the SessionStateKeeper. If so, use the registered SessionState.
+                // - Otherwise, check if default SessionState is still alive. If so, use the default SessionState.
+                // - Otherwise, the 'SessionStateInternal' property will be set to null, and the current default runspace will be used.
```

Comment:
I understand that, but I feel the comment is wrong.  I believe it should say: "Otherwise, the 'SessionStateInternal' property will be set to null and the current runspace will be used."

Not "the current default runspace will be used" since there is no default runspace...